File: Array/2099.Find_Subsequence_of_Length_K_With_the_Largest_Sum/Python/2099.Find_Subsequence_of_Length_K_With_the_Largest_Sum.py
class Solution(object):
    def maxSubsequence(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        #===========================================#
        # Heap/Priorit-queue based traversal method #
        #===========================================#

        ############
        #Initialize
        ##### Length of nums array #####
        len_nums = len(nums)

        ##### Record heap/priority-queue #####
        record_pq = []

        heapify(record_pq)

        ##### Record array #####
        record_arry = [((-1) * ((10 **6))) for _ in range(len_nums)]

        ##### Result array #####
        res_arry = []


        #########################################
        #Heap/Priorit-queue based loop traversal
        ##### Step 1: Record indexed-value with heap/priority-queue #####
        for nums_idx in range(len_nums):
            heappush(record_pq, [nums[nums_idx], nums_idx]) #Keep updating/recording

            ##### Check if the current size is larger than k-th or not #####
            if (len(record_pq) > k):
                heappop(record_pq) #Keep updating/popped
            else:
                pass

        ##### Step 2: Looped-traversal with recorded heap/priority-queue #####
        while (record_pq):
            record_arry[(record_pq[0])[1]] = (record_pq[0])[0] #Keep updating/overwriting

            heappop(record_pq) #Keep updating/popped

        ##### Step 3: Post-process/Final completions #####
        for nums_idx in range(len_nums):

            ##### Check if the current indexed-value matched conditions or not #####
            if (record_arry[nums_idx] != ((-1) * ((10 **6)))):
                res_arry.append(record_arry[nums_idx]) #Keep updating/recording
            else:
                pass

        return res_arry

File: Array/2099.Find_Subsequence_of_Length_K_With_the_Largest_Sum/C++/2099.Find_Subsequence_of_Length_K_With_the_Largest_Sum.cpp
class Solution {
public:
    vector<int> maxSubsequence(vector<int> & nums, int k) 
    {
        //====================================//
        // Input type:                        //
        //   - vector<int> ref. (nums vector) //
        //   - int (k-th quantities)          //
        // Return type:                       //
        //   - vector<int> (result vector)    //
        //====================================//

        /*Initialize*/
        ///// Length of nums vector /////
        int len_nums = nums.size();

        ///// Record heap, priority-queue /////
        priority_queue<pair<int, int>> record_pq;

        ///// Record vector /////
        vector<int> record_vect(len_nums, ((-2) * (1e5)));

        ///// Result vector /////
        vector<int> res_vect;


        /*Heap, Priority-queue based loop traversal*/
        ///// Step 1: Record indexed-value with heap, priority-queue /////
        for (int nums_idx = 0; (nums_idx < len_nums); (nums_idx++)) //Whole
        {
            record_pq.push(make_pair(((-1) * nums[nums_idx]), nums_idx)); //Keep updating, recording

            ///// Check if the current size is larger than k-th or not /////
            if (record_pq.size() > k) { record_pq.pop(); } //Keep updating, popped
            else { ; }
             
        } //Whole

        ///// Step 2: Looped-traversal with recorded heap, priority-queue /////
        while (! record_pq.empty()) //Remains
        {
            record_vect[(record_pq.top().second)] = ((-1) * (record_pq.top()).first); //Keep updating, overwriting

            record_pq.pop(); //Keep updating, popped

        } //Remains

        ///// Step 3: Post-process, Final completions /////
        for (int nums_idx = 0; (nums_idx < len_nums); (nums_idx++)) //Whole
        {
            ///// Check if the current indexed-value matched conditions or not /////
            if (record_vect[nums_idx] != ((-2) * (1e5))) { res_vect.emplace_back(record_vect[nums_idx]); } //Keep updating, recording
            else { ; }

        } //Whole

        return res_vect;
    }
};
